proposals/code-signing-for-homebrew.md

```diff
@@ -0,0 +1,101 @@
+
+To verify a Sigstore signature, a particular `brew` client needs the following:
+
+- A trust proposition, i.e. a reason to trust the identity producing the signature. For Homebrew, this trust proposition comes from the tap's own identity: the act of performing `brew tap foo/bar` establishes that Sigstore identities tied to `github.com/foo/bar` can be trusted to sign for bottles in that tap. This applies to implicit `brew tap` invocations as well, e.g. `brew install foo/bar/baz` asserts that the user expects a signature on `baz` from the `foo/bar` GitHub identity.
+- A Sigstore client library capable of verifying Sigstore signatures.
+
+Client-side verification of bottles requires `brew` to know which bottles are actually signed. Clients should use the "flag day" metadata above to establish this: `brew install` should refuse to install any unsigned bottle from a tap that both had a flag day _and_ whose flag day is before the local system time. 
+
+The high-level flow for bottle signature verification, at tap time:
+
+- A user performs either an explicit tap (`brew tap foo/bar`) or an implicit tap (`brew install foo/bar/baz`), and the local `brew` client establishes a trust proposition based on that action ("bottles from `foo/bar` must be signed by the `foo/bar` identity").
+- The local `brew` client looks for the "flag day" metadata and, if present and before the local system time, enforces signature verification for all bottles referenced by the tap.
+
+And at `brew install` time:
+
+- For each bottle resolved during a `brew install` invocation, the bottle must have an accompanying signature. If the accompanying signature is not present, then `brew install` must fail.
+- For each bottle and signature, `brew install` must internally reconstruct the _attestation_ described in the signing process above. This is then used as the cleartext input for signature verification. This input is then passed into the underlying Sigstore client along with its associated signature; the Sigstore client is responsible for performing both the signature verification and its associated identity verification.
+- If the underlying signature verification fails, then `brew install` must fail.
+
+Like with previous Homebrew features, an experimental feature-flagged variant of signature verification may be merged and enabled before it is enabled by default.
+
+## Full source and build provenance with Homebrew and Sigstore
+
+Bottle signing and subsequent verification unlocks a potential subsequent improvement: because Homebrew performs source builds to produce bottles, Homebrew could also verify signatures for those source artifacts. That signature could then be _countersigned_ for by Homebrew's own bottle signatures, effectively producing a strong cryptographic association from the original source artifact (produced by a third-party upstream) to the Homebrew-produced binary artifact.
```

Comment:
Could this use the GitHub Actions generator for SLSA provenance (for build provenance, source provenance is a WIP track for SLSA)? - https://github.com/slsa-framework/slsa-github-generator

Reply:
Yes! I think that's probably exactly what we want to do 🙂 -- I think the only outstanding big question there is around discoverability (i.e., how should Homebrew go from "here's my source archive" to "I know that there's a SLSA provenance available for this source"?)

Reply:
Maybe @steiza can chime in on this specific topic?

Reply:
When we built the attestation store for npm, we had in mind that we could extend it so GitHub repos could also store and fetch attestation documents. We're thinking of exposing a public API where you supply the repo and artifact hash, and get back the associated attestation document.

Reply:
> I suppose one issue is that it's not yet clear how sigstore signed attestations work in bundles?

There's some unfortunate terminology overlap here: the "bundles" in this proposal are Sigstore bundles, which include the signature, signing certificate, transparency log entries, etc. 

The signature in turn is performed over the (in-toto formatted) attestation, which is also committed to the transparency log (since Rekor has a dedicated type for in-toto attestations). 

In other words: the top level "unit" here is a Sigstore bundle, which then contains the signed "thing". That "thing" could itself be the bottle attestation (what I was planning) or an in-toto bundle, but that latter one then means multiple layers of signing and I need to think about that some more 🙂

Reply:
Yeah, avoiding multiple levels of signing would be good.  FWIW @haydentherapper and I had discussed that it might be worthwhile to figure out either how to

1. Put a _Sigstore_ bundle in an in-toto bundle (e.g. can it just be dropped in a jsonl file and work for folks that expect it), or
2. Figure out how to get Sigstore to embed the signing information it needs inside the DSSE itself instead of wrapping it.

My intuition is that 1 should be pretty easy, but that 2 is the 'right' solution but will take much longer and won't be backwards compatible.

Reply:
Yeah, (1) seems easier!

I'm a little bit less familiar with in-toto bundles, but from a quick scan they're JSON lines, with each JSON payload being an in-toto `Envelope`. Those envelopes are then signed, so would this be double-signing again? Or am I missing something?

Reply:
Re @woodruffw 

There doesn't need to be double-signing, what I'm thinking for option 1 is that we can potentially rely on ["Consumers MUST ignore unrecognized lines"](https://github.com/in-toto/attestation/blob/main/spec/v1/bundle.md#:~:text=Consumers%20MUST%20ignore%20unrecognized%20lines.) and just format each _Sigstore_ bundle into a single line and add to the _in-toto_ bundle without doing any additional work.  So, a bit of a hack.

Re @adityasaky yes! That seems like the ideal solution (corresponding to (2) above?).  Does anyone think that can happen quickly enough to rely on it for this use case or if we should do (1) instead?

Reply:
I can submit a proposal to the DSSE spec based on the conversations in that issue but I'm not 100% certain about how its adoption will progress in sigstore. @haydentherapper thoughts?

Reply:
https://github.com/sigstore/sig-clients/issues/8 to discuss with the sigstore community a bundle format

Reply:
> Is the idea future consumers might understanding it, and could do more useful things with it?

Yes exactly.  Consumers would need to support multiple different types of lines in the in-toto bundles, but that seems doable (even if it's inelegant).

Reply:
Makes sense to me. I'm happy to bundle the actual signing outputs here in whatever format makes things easiest for additional future users; if that means rolling it into another surrounding bundle, that's fine!

Reply:
Resolved per #27.